src/compute-client/src/controller/instance.rs

```diff
@@ -306,10 +306,51 @@ where
     /// This method removes the replica from the orchestrator and should only be called if the
     /// replica should be permanently removed.
     pub async fn remove_replica(&mut self, id: ReplicaId) -> Result<(), ComputeError> {
+        if !self.compute.replicas.contains_key(&id) {
+            return Err(ComputeError::ReplicaMissing(id));
+        };
+
         if let Err(e) = self.compute.replicas[&id].send(ComputeCommand::DropInstance) {
             tracing::warn!("Could not send DropInstance to replica {:?}: {}", &id, &e)
         }
 
+        // If the replica is managed we have to remove it from the orchestrator. We spawn
+        // a background task that waits until the termination of the message handler task and
+        // then removes it from the orchestrator.
+        if matches!(
+            self.compute.replicas[&id].location,
+            ComputeReplicaLocation::Managed { .. }
+        ) {
+            let replica_task = self
+                .compute
+                .replicas
+                .get_mut(&id)
+                .unwrap()
+                .replica_task
+                .take()
+                .unwrap();
+            let instance_id = self.compute.instance_id.clone();
+            let orchestrator = self.compute.orchestrator.clone();
+            let id = id.clone();
+            mz_ore::task::spawn(|| format!("drop-replica-{id}"), async move {
+                // Ensure the active-replication-replica task is terminated before removing the service
+                // from the orchestrator. This await guarantees the ensure call has happened before
+                // we remove the replica from the orchestrator.
+                replica_task.abort();
+                let join_result = replica_task.await;
+                tracing::debug!("Replica task joined: {:?}", join_result);
```

Comment:
I assume the waiting is done to avoid the scenario where we first drop the replica here and then the `replica_task` wakes up and calls the orchestrator to add it again?

This solution seems fine, but I wonder if it wouldn't be simpler to have a serializing `orchestrator_task` instead. One per replica or a single one for all replicas, not sure if that is important. The task would listen to `Ensure` and `Drop` commands on a channel and execute them in the order received. Then in `add_replica` and `remove_replica` we would just send commands on the channel and that's it. The main benefit of that approach would be that the code would be structured in a clearer way, so the potential for bugs is hopefully lower as well.

Reply:
I'm generally pro OrchestratorTask, but the problem is that ReplicaTask needs the result of add_replica to construct the CommunicationConfig. The OrchestratorTask could return a oneshot channel that we then pass into the ReplicaTask, which would then await the operation to complete. But that feels like an overcomplicated way of simply calling the orchestrator in the ReplicaTask...

Reply:
Fair enough! I still think an orchestrator task might lead to clearer code, but I have only intuition to back this up and since it's not a trivial change, it's probably not worth the effort to try it out.